cgen error: mismatch when the default expression or block has optional

OS: macos, macOS, 12.6, 21G115
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
CC version: Apple clang version 14.0.0 (clang-1400.0.29.102)

getwd: /usr/local/v
vmodules: /Users/shove/.vmodules
vroot: /usr/local/v
vexe: /usr/local/v/v
vexe mtime: 2022-10-12 11:05:35
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 34d115d.c590c82

Git version: git version 2.37.0 (Apple Git-136)
Git vroot status: weekly.2022.40-54-gc590c825-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 46662e20


**What did you do?**

```v
struct Foo {}

fn bar() ?&Foo {
	return none
}

fn main() {
	// _ := bar() or { Foo{} }  // this OK.
	bar() or { Foo{} }
}
```

**What did you expect to see?**

nothing
 
**What did you see instead?**

```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/a.11985886069740129955.tmp.c:12048:27: error: assigning to 'main__Foo *' (aka 'struct main__Foo *') from incompatible type 'main__Foo' (aka 'struct main__Foo'); take the address with &
                *(main__Foo**) _t1.data = ((main__Foo){EMPTY_STRUCT_INITIALIZATION});
                                        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                          &(                                        )
/tmp/v_501/a.11985886069740129955.tmp.c:12051:4: warning: expression result unused [-Wunused-value]
  (*(main__Foo**)_t1.data);
   ^~~~~~~~~~~~~~~~~~~~~~
2 warnings and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1070,7 +1070,9 @@ fn (mut c Checker) check_or_last_stmt(stmt ast.Stmt, ret_type ast.Type, expr_ret
 					return
 				}
 				if c.check_types(stmt.typ, expr_return_type) {
-					return
+					if stmt.typ.is_ptr() == expr_return_type.is_ptr() {
+						return
+					}
 				}
 				// opt_returning_string() or { ... 123 }
 				type_name := c.table.type_to_str(stmt.typ)

--- a/vlib/v/checker/tests/optional_or_block_mismatch.out
+++ b/vlib/v/checker/tests/optional_or_block_mismatch.out
@@ -4,4 +4,10 @@ vlib/v/checker/tests/optional_or_block_mismatch.vv:10:18: error: wrong return ty
    10 |     x := foo() or { Bar{} }
       |                     ~~~~~
    11 |     println(x)
-   12 | }
+   12 |
+vlib/v/checker/tests/optional_or_block_mismatch.vv:13:13: error: the default expression type in the `or` block should be `&Bar`, instead you gave a value of type `Bar`
+   11 |     println(x)
+   12 | 
+   13 |     foo() or { Bar{} }
+      |                ~~~~~
+   14 | }

--- a/vlib/v/checker/tests/optional_or_block_mismatch.vv
+++ b/vlib/v/checker/tests/optional_or_block_mismatch.vv
@@ -9,4 +9,6 @@ fn foo() ?&Bar {
 fn main() {
 	x := foo() or { Bar{} }
 	println(x)
+
+	foo() or { Bar{} }
 }

